### DISCUSSION TITLE
How to use I2C?

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#742] [Q&A] [answered]
Hello,

How do I instantiate I2C ports in my component? I am using Fprime v1.5.3. Right now, I have this code in my Component:

```
<!-- Import ports for i2c driver -->
<import_port_type>Drv/I2cDriverPorts/I2cPortAi.xml</import_port_type>
<ports>
        <port name="I2CWrite" data_type="Drv::I2c"  kind="guarded_input">
        </port>
        <port name="I2CRead" data_type="Drv::I2c"  kind="guarded_input">
        </port>
 </ports>
```

and my cpp-template creates two functions for I2Cwrite_handle an I2CRead_Handler. Why does it create these two handlers? I thought it would be the same as serial driver where there is one handle for read but you can write in any other functions.

**Comment by timcanham** [comment, marked as answer]
@sheershak-agarwal You should make the port kind `output`, then you can call the correct member function in the base class. Since I2C is synchronous, it makes sense that the component sending the transaction would have both operations as output ports.

**Reply by sheershak-agarwal** [reply]
I made read and write ports both as output. There are no functions created now. So now I have to call the read and write functions to get and send the data. Is there a example component which shows how to read and write from I2C?

**Reply by rawinza555** [reply]
There is a RPI demo that did something similar to this with uart and spi. I bet for i2c it should be somewhat similar functionality wise.

**Reply by LeStarch** [reply]
Yes the SPI port in the RPI demo works similarly.  Just remember SPI is one transaction (read and write) and I2C you have a choice of read, write, write then read.

**Reply by sheershak-agarwal** [reply]
Thanks

**Reply by rawinza555** [reply]
@LeStarch I checked the I2c writeread ports and see that it has a return value. I thought the port is supposed to be in only one direction. 

So if I want to use the return value I could have just do something like 
`a = I2cWriteRead()_out `
in my componentimpl.cpp right?
